C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp

```diff
@@ -0,0 +1,87 @@
+// https://en.wikipedia.org/wiki/Knuth–Morris–Pratt_algorithm
+
+// Time Complexity:-O(N+M).
+// Space Complexity:-O(M).
+
+#include <iostream>
+#include <vector>
+#include <string>
+
+using namespace std;
+
+void computeLPS(string &, vector<int> &);
+void searchByKMP(string &, string &);
+
+int main()
+{
+    string text = "abacaaabaccabacabaabb";
+    string patt = "abacab";
+
+    searchByKMP(text, patt);
+    return 0;
+}
+// Prints starting index where patt is found in the string text.
+
+void searchByKMP(string &text, string &patt)
+{
+    int N = text.size(), M = patt.size();
+
+    vector<int> lps(M);
+
+    computeLPS(patt, lps);
+
+    int i = 0, j = 0;
+
+    while (i < N)
+    {
+        if (patt[j] == text[i])
+        {
+            i++;
+            j++;
+        }
+
+        if (j == M)
+        {
+            cout << "Pattern found at index " << i - j << endl;
+            j = lps[j - 1];
+        }
+        else if (i < N && patt[j] != text[i])
+        {
+            if (j == 0)
+                i++;
+            else
+                j = lps[j - 1];
+        }
+    }
+}
+
+// Function to Compute LPS.
+// LPS = Longest Prefix which also a suffix.
```

Comment:
Use proper comments to explain your code.

Reply:
@atarax665 Comments Updated.